pyrolite/mineral/normative.py

```diff
@@ -301,308 +302,661 @@ def CIPW_norm(df, Fe_correction=None, adjust_all=False):
+
+    df['Di_'] = np.where((df['D5'] < df['Di_p']), df['Di_p'] - df['D5'], 0).T
+
+    df['D6'] = df['D5'] - df['Di_p']
+
+    df['Cs'] = np.where((df['deficit']), df['Cs_'], df['Cs']).T
+    df['Ol'] = np.where((df['deficit']), df['Ol_'], df['Ol']).T
+    df['Di'] = np.where((df['deficit']), df['Di_'], df['Di_p']).T
+
+    df['deficit'] = df['D6'] > 0
+
+    # Normative Kaliophilite / Leucite
+    df['Kp'] = np.where((df['Lc'] >= df['D6'] / 2), df['D6'] / 2, df['Lc']).T
+
+    df['Lc_'] = np.where((df['Lc'] >= df['D6'] / 2), df['Lc'] - df['D6'] / 2, 0).T
+
+    df['Kp'] = np.where((df['deficit']), df['Kp'], 0).T
+    df['Lc'] = np.where((df['deficit']), df['Lc_'], df['Lc']).T
+
+    df['DEFSIO2'] = np.where((df['Lc'] < df['D6'] / 2) & (df['deficit']), df['D6'] - 2 * df['Kp'], 0).T
+
+    # Allocate definite mineral proportions
+    ## Subdivide Hypersthene, Diopside and Olivine into Mg- and Fe- varieties
+
+    df['Fe-Hy'] = df['Hy'] * df['FeO_ratio']
+    df['Fe-Di'] = df['Di'] * df['FeO_ratio']
+    df['Fe-Ol'] = df['Ol'] * df['FeO_ratio']
+
+    df['Mg-Hy'] = df['Hy'] * df['MgO_ratio']
+    df['Mg-Di'] = df['Di'] * df['MgO_ratio']
+    df['Mg-Ol'] = df['Ol'] * df['MgO_ratio']
+
+    mineral_proportions = pd.DataFrame()
+    mineral_pct_mm = pd.DataFrame()
+    FREE = pd.DataFrame()
+
+    FREE['FREEO_12b'] = (
+                                1 + ((0.1) * ((mineral_molecular_weights['CaF2-Ap'] / 328.86918) - 1))
+                        ) * element_AW['O'] * df['FREEO_12b']
+    FREE['FREEO_12c'] = (
+                                1 + ((0.1) * (df['CaF2-Ap'] / df['Ap']) * (
+                                    (mineral_molecular_weights['CaF2-Ap'] / 328.86918) - 1))
+                        ) * element_AW['O'] * df['FREEO_12c']
+
+    FREE['FREEO_13'] = (
+                               1 + ((pt.formula('CaO').mass // 56.0774) - 1)
+                       ) * element_AW['O'] * df['FREEO_13']
+
+    FREE['FREEO_14'] = (
+                               1 + (0.5 * ((pt.formula('Na2O').mass / 61.9789) - 1))
+                       ) * element_AW['O'] * df['FREEO_14']
+
+    FREE['FREEO_16'] = (
+                               1 + ((pt.formula('FeO').mass // 71.8444) - 1)
+                       ) * element_AW['O'] * df['FREEO_16']
+
+    FREE['O'] = FREE[['FREEO_12b', 'FREEO_12c', 'FREEO_13', 'FREEO_14', 'FREEO_16']].sum(axis=1)
+
+    FREE['CO2'] = df['FREECO2'] * 44.0095
+
+    FREE['P2O5'] = df['FREE_P2O5'] * 141.94452
+    FREE['F'] = df['FREE_F'] * 18.9984032
+    FREE['Cl'] = df['FREE_Cl'] * 35.4527
+    FREE['SO3'] = df['FREE_SO3'] * 80.0642
+    FREE['S'] = df['FREE_S'] * 32.066
+    FREE['Cr2O3'] = df['FREE_CR2O3'] * 151.990
+
+    FREE['OXIDES'] = FREE[['P2O5', 'F', 'Cl', 'SO3', 'S', 'Cr2O3']].sum(axis=1)
+    FREE['DEFSIO2'] = df['DEFSIO2'] * 60.0843
+    FREE.drop(['P2O5', 'F', 'Cl', 'SO3', 'S', 'Cr2O3'], axis=1, inplace=True)
+
+    not_subdivided = ['Hy', 'Di', 'Ol']
+    mineral_codes = {
```

Comment:
If we use the 'minerals' dictionary I commented at the start, this data can live in there too.

Reply:
ive got rid of this - by default it will return all minerals